### PR TITLE
Use FP32 for smoothing

### PR DESCRIPTION
**Pull request by MalachiTimothyPhillips**
Fuse collocation inside other kernels for FDM
Avoid allocating extraneous work arrays
Overlap fusedFDM kernel
For k=1, fuse a few scaledAddKernels inside the chebyshev smoother for increased performance

### CODE REVIEW

File: src/libP/solvers/elliptic/okl/ellipticSerialAxHex3D.c [hunk 1]
```diff
@@ -118,6 +118,100 @@ void ellipticAxHex3D(const dlong & Nelements,
         }
   }
 }
+extern "C"
+void ellipticAxFloatHex3D(const dlong & Nelements,
```
Comment (stgeke) [inline src/libP/solvers/elliptic/okl/ellipticSerialAxHex3D.c:122]: I don't think we need an extra float version. Just compile the kernel with a different dfloat type.

File: src/libP/solvers/elliptic/src/ellipticMultiGridLevel.c [hunk 2]
```diff
@@ -64,12 +64,16 @@ void MGLevel::prolongate(occa::memory o_x, occa::memory o_Px)
 
 void MGLevel::smooth(occa::memory o_rhs, occa::memory o_x, bool x_is_zero)
 {
+  elliptic->copyDfloatToPfloatKernel(Nrows, o_float_x, o_x);
```
Comment (stgeke) [inline src/libP/solvers/elliptic/src/ellipticMultiGridLevel.c:67]: Do not copy if Dfloat == Pfloat

File: src/libP/solvers/elliptic/src/ellipticMultiGridLevel.c [hunk 3]
```diff
@@ -64,12 +64,16 @@ void MGLevel::prolongate(occa::memory o_x, occa::memory o_Px)
...
   if (stype == RICHARDSON)
-    this->smoothRichardson(o_rhs, o_x, x_is_zero);
+    this->smoothRichardson(o_float_rhs, o_float_x, x_is_zero);
   else if (stype == CHEBYSHEV)
-    this->smoothChebyshev(o_rhs, o_x, x_is_zero);
+    this->smoothChebyshev(o_float_rhs, o_float_x, x_is_zero);
   else if (stype == SCHWARZ)
-    this->smoothSchwarz(o_rhs, o_x, x_is_zero);
+    this->smoothSchwarz(o_float_rhs, o_float_x, x_is_zero);
+  elliptic->copyPfloatToDfloatKernel(Nrows, o_float_x, o_x);
```
Comment (stgeke) [inline src/libP/solvers/elliptic/src/ellipticMultiGridLevel.c:75]: Do not copy if Dfloat == Pfloat

File: src/libP/solvers/elliptic/src/ellipticMultiGridLevel.c [hunk 4]
```diff
@@ -136,51 +174,42 @@ void MGLevel::smoothChebyshev (occa::memory &o_r, occa::memory &o_x, bool xIsZer
     this->smoother(o_r, o_res, xIsZero);
 
     //d = invTheta*res
-    elliptic->scaledAddKernel(Nrows, invTheta, o_res, zero, o_d);
+    elliptic->scaledAddFloatKernel(Nrows, invTheta, o_res, zero, o_d);
```
Comment (stgeke) [inline src/libP/solvers/elliptic/src/ellipticMultiGridLevel.c:177]: Rename to scaledAddPfloatKernel

File: src/libP/solvers/elliptic/okl/ellipticAxHex3D.okl [hunk 5]
```diff
@@ -424,13 +423,13 @@
...
 @kernel void ellipticPartialAxHex3D_v0(const dlong Nelements,
                                        @restrict const dlong*  elementList,
-                                       @restrict const dfloat*  ggeo,
-                                       @restrict const dfloat*  D,
-                                       @restrict const dfloat*  S,
-                                       @restrict const dfloat*  MM,
-                                       const dfloat lambda,
-                                       @restrict const dfloat*  q,
-                                       @restrict dfloat*  Aq)
+                                       @restrict const pfloat*  ggeo,
```
Comment (stgeke) [inline src/libP/solvers/elliptic/okl/ellipticAxHex3D.okl:426]: Why was this changed? I think this should be dfloat.

File: src/libP/include/types.h [hunk 6]
```diff
@@ -39,7 +39,7 @@ SOFTWARE.
 #endif
 
 //smoother float data type
-#if 1
+#if 0
```
Comment (stgeke) [inline src/libP/include/types.h:42]: Use Pfloat = float by default

File: src/libP/include/mesh.h [hunk 7]
```diff
@@ -350,9 +350,10 @@ typedef struct {
 
   occa::memory o_q, o_rhsq, o_resq, o_fQM, o_fQP;
 
-  occa::memory o_Dr, o_Ds, o_Dt, o_LIFT, o_MM;
+  occa::memory o_Dr, o_Ds, o_Dt, o_LIFT, o_MM, o_float_MM;
```
Comment (stgeke) [inline src/libP/include/mesh.h:353]: Rename variable like this to o_MMPfloat

File: src/libP/okl/scaledAdd.okl [hunk 8]
```diff
@@ -24,6 +24,41 @@ SOFTWARE.
...
+                      @restrict dfloat * res,
+                      @restrict dfloat * d,
+                      @restrict dfloat * x){
+  
+  for(dlong n=0;n<N;++n;@tile(256,@outer,@inner)){
+    if(n<N){
+      const dfloat value = -1.0*Ad[n] + 1.0*res[n];
+      res[n] = value;
+      const dfloat d_value = rhoDivDelta * value + rho_np1 * rho_n * d[n];
+      d[n] = d_value;
```
Comment (stgeke) [inline src/libP/okl/scaledAdd.okl:57]: I don't think we need to write out d[]

File: src/libP/solvers/elliptic/src/ellipticOperator.c [hunk 9]
```diff
@@ -52,6 +52,20 @@ void ellipticAx(elliptic_t* elliptic,
...
+    valid &= continuous;
+    if(!strstr(precision, dfloatString)){
+      valid &= !elliptic->var_coeff;
+      valid &= !elliptic->blockSolver;
+      if(!serial){
+        valid &= mapType == 0;
+        valid &= integrationType == 0;
+      }
+    }
+    if(!valid) exit(1);
```
Comment (stgeke) [inline src/libP/solvers/elliptic/src/ellipticOperator.c:66]: Print an error message instead of simply bailing out

File: src/libP/okl/scaledAdd.okl (hunk 8 above)
Comment (MalachiTimothyPhillips) [reply src/libP/okl/scaledAdd.okl:57]: I found that res also doesn't need to be written.

**Review by stgeke — APPROVED** [review]

**Comment by stgeke** [comment]
Replaced by #142
